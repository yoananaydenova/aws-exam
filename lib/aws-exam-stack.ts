import * as cdk from "aws-cdk-lib";
import {
  AttributeType,
  BillingMode,
  StreamViewType,
  Table,
} from "aws-cdk-lib/aws-dynamodb";
import {
  Code,
  FilterCriteria,
  FilterRule,
  Runtime,
  StartingPosition,
} from "aws-cdk-lib/aws-lambda";
import { DynamoEventSource } from "aws-cdk-lib/aws-lambda-event-sources";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Subscription, SubscriptionProtocol, Topic } from "aws-cdk-lib/aws-sns";
import { Construct } from "constructs";

export class AwsExamStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const fileBucket = new Bucket(this, "FileUploadBucket", {
      lifecycleRules: [{ expiration: cdk.Duration.minutes(30) }],
    });

    const fileTable = new Table(this, "FileMetadata", {
      partitionKey: { name: "fileId", type: AttributeType.STRING },
      sortKey: { name: "fileExtension", type: AttributeType.STRING },
      billingMode: BillingMode.PAY_PER_REQUEST,
      timeToLiveAttribute: "ttl",
      stream: StreamViewType.NEW_AND_OLD_IMAGES,
    });

    const notifyTopic = new Topic(this, "FileUploadNotifications");

    new Subscription(this, "ErrorSubscription", {
      topic: notifyTopic,
      protocol: SubscriptionProtocol.EMAIL,
      endpoint: "hristo.zhelev@yahoo.com",
    });

    const fileProcessingFunction = new NodejsFunction(
      this,
      "fileProcessingFunction",
      {
        runtime: Runtime.NODEJS_20_X,
        handler: "handler",
        entry: `${__dirname}/../src/processFunction.ts`,
        environment: {
          BUCKET: fileBucket.bucketName,
          TABLE: fileTable.tableName,
          SNS_TOPIC: notifyTopic.topicArn,
        },
      }
    );

    const cleanupFunction = new NodejsFunction(this, "cleanupFunction", {
      runtime: Runtime.NODEJS_20_X,
      handler: "handler",
      entry: `${__dirname}/../src/cleanupFunction.ts`,
      // memorySize: 128,
      environment: {
        TABLE: fileTable.tableName,
        SNS_TOPIC: notifyTopic.topicArn,
      },
    });

    cleanupFunction.addEventSource(
      new DynamoEventSource(fileTable, {
        startingPosition: StartingPosition.LATEST,
        batchSize: 5,
        filters: [
          FilterCriteria.filter({
            eventName: FilterRule.isEqual("REMOVE"),
          }),
        ],
      })
    );

    fileBucket.grantReadWrite(fileProcessingFunction);
    fileTable.grantReadWriteData(fileProcessingFunction);
    notifyTopic.grantPublish(fileProcessingFunction);

    new cdk.CfnOutput(this, "Stack", {
      value: "Result: ",
    });
  }
}
