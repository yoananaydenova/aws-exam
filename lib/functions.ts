import { Duration } from "aws-cdk-lib";
import { Architecture, Runtime } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";

interface BaseFunctionProps {
  tableName: string;
  handlerLocation: string;
}

export class BaseFunction extends NodejsFunction {
  constructor(scope: Construct, id: string, props: BaseFunctionProps) {
    super(scope, id, {
      ...props,
      runtime: Runtime.NODEJS_20_X,
      handler: "handler",
      entry: `${__dirname}/../src/${props.handlerLocation}.ts`,
      timeout: Duration.seconds(10),
      architecture: Architecture.ARM_64,
      environment: {
        // enviroment variables
        TABLE_NAME: props.tableName,
      },
    });
  }
}