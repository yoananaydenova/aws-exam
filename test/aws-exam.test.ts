import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as AwsExam from "../lib/aws-exam-stack";
import "jest-cdk-snapshot";

test("Stack is changed", () => {
  const app = new cdk.App();
  // WHEN
  const stack = new AwsExam.AwsExamStack(app, "TestStack");
  // THEN
  expect(stack).toMatchCdkSnapshot();
});
