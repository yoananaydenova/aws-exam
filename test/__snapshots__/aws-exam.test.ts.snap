// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack is changed 1`] = `{}`;
